### PR TITLE
Fix curl creating download dir in 750 mode

### PR DESCRIPTION
**Pull request by moritz-h**
This suggests to fix the download directory being created with 750 mode on unix like systems.
Currently, the downloads directory is not readable by other users. This is inconvenient, e.g., for debugging CI systems, when the build directory is not fully readable. Further, this is an inconsistency to all other directories created by vcpkg (buildtrees, packages, installed, ...). All other directories are created with the default mkdir mode.

I traced the problem down to curl being used with the `--create-dirs` option to create all download directories. Curl always uses mode 750 to create directories as written [in the docs](https://curl.se/docs/manpage.html#--create-dirs).

This PR fixes the problem by creating the download directory in advance. I found an other usage of `--create-dirs` in the bulk download code: https://github.com/microsoft/vcpkg-tool/blob/2024-06-10/src/vcpkg/base/downloads.cpp#L481
But it seems file system access is not already available there, so maybe waiting for feedback if that solution is valid at all before trying to fix this there.

### CODE REVIEW

**Comment by dg0yt** [comment]
> directory being created with 750 mode on unix like systems.

This seems surprising given that there is `umask`. That curl code is at least 13 years old. I wonder if it is intentional.

**Comment by moritz-h** [comment]
> This seems surprising given that there is `umask`. That curl code is at least 13 years old. I wonder if it is intentional.

It seems intentional as it was documented resulting from this discussion: https://github.com/curl/curl/issues/4766
In that discussion it was also mentioned that umask is respected. Only 750 is used as default before applying umask.

**Comment by dg0yt** [comment]
Hm, this discussion isn't entirely neutral...

The point isn't that umask isn't applied. The point is that curl starts with a more restrictive mode than expected (from common practice), and this is surprising. The world part of umask is rendered useless. Users must explicit call `chmod` to relax  world restrictions.

(One might see this is as a security feature... but calling `chmod` is a security trap: No umask applied. Easy to make things world writable.)

**Comment by moritz-h** [comment]
> a more restrictive mode than expected (from common practice), and this is surprising.

That is exactly the point for this PR. We probably have the same opinion on this.
I think its strange that the download folder has not default folder permission applied compared to a mkdir call on that system. The curl discussion likely was not optimal especially from the request side. However, it was made clear from curl that 750 is the intended behaviour. Therefore, it may be fixed in vcpkg.
I also do not like to play around with file permissions after the curl call, especially as you may need to track what part of the path existed before and which folders are newly created. Thats why I'm suggesting to just create the folders in advance here.

File: src/vcpkg/base/downloads.cpp [hunk 1]
```diff
@@ -786,6 +786,10 @@ namespace vcpkg
             }
         }
 #endif
+        // Create directory in advance, otherwise curl will create it in 750 mode on unix style file systems.
+        const auto dir = download_path_part_path.parent_path();
```
Comment (BillyONeal) [inline src/vcpkg/base/downloads.cpp:790]: I'm not sure this is the right thing, there are a lot of edge cases where just chopping off the last path element will do the wrong thing. Consider something like `a/b/../out.exe`.

**Review by BillyONeal — COMMENTED** [review]
At a minimum this needs a test and to remove the `--create-dirs` part. The purpose of `--create-dirs` is to deal with racy behavior: The right behavior isn't to try to create directories before doing a thing, it's to try to create the file directly, and only if that fails fall back to trying to create directories and the hope is that curl is able to do such behavior.
